### ISSUE TITLE
Improvements to the delivery report using the opaque

### ISSUE THREAD

**Issue by mimaison** [Blizzard/node-rdkafka#100]
The addition of the opaque is great (performance/flexibility) but:

1. In order to have the value in the delivery report, it forces the user to make a "clunky" API call.
For example:
```
var value = new Buffer('...');
var key = 'key';
producer.produce('topic', null, value, key, {value: value, key: key});
```

It feels a lot more natural to automatically get the opaque populated in case the user has not passed in a custom value. And this does not cause any performance penalty as it reuses the opaque mechanism.

2. The key member of the delivery report object still suffers from the null/empty issue. It can be correctly retrieve for free from the opaque as described above, hence removing the need to have native code handling the key. 

Although people generally consider to be small, in Kafka it is actually the same datatype as the payload and can be arbitrary big. So reusing the opaque to pass it along the payload make sense.

**Comment by InfinitiesLoop** [comment]
I can understand the convenience factor of the key/value being a default opaque value. But while it does not seem it would have a big performance penalty, it isn't free either. This trades one scenario for another -- someone might want delivery reports and do not need an opaque value, in which case they'd have to pass in null explicitly (to reduce the unnecessary shuffling), which is also awkward.

Having no default opaque is truer to the goal of this library; to be a node wrapper for librdkafka, and librdkafka does not have this behavior.

**Comment by edoardocomar** [comment]
@InfinitiesLoop  
The tradeoff is not only convenience/speed. 
Having application code simpler means less API to maintain compatibility with.

librdkafka assumes the behavior is to have a full delivery report (@edenhill mentioned that dr_msg_cb is the setting maintained going forward)

Are you really interested in having a delivery report without the original message payload?
Because not having the delivery report at all would of course eliminate any performance penalty (we suggest to have the default opaque only if you choose to have the cb.

I am skeptic passing around a pointer will incur a noticeable penalty. With all the rest of the binding operations, and having got rid of dealing with the key. 

May be worth running a quick perf test.

**Comment by edoardocomar** [comment]
We could also choose to have the convenient opaque only if you have set `dr_msg_cb` and not if you have just set `dr_cb`.

**Comment by edoardocomar** [comment]
I ran `bench/producer-raw-rdkafka.js` with the default settings, on my mac with a local kafka.
3 runs with each codebase (alternating master/patched)

master (no automagic opaque)
total: 9999999 messages over 348897 ms
28661 messages / second
total: 9999999 messages over 349916 ms
28578 messages / second
total: 9999999 messages over 356138 ms
28079 messages / second

patched (automagic opaque)
total: 9999999 messages over 352747 ms
28348 messages / second
total: 9999999 messages over 351743 ms
28429 messages / second
total: 9999999 messages over 356041 ms
28086 messages / second 

so the difference is ~0.5% 

@InfinitiesLoop 
if you are really this concerned this difference matters and can't bear passing a null opaque,
I propose using the automagic opaque only with `dr_msg_cb` set, not with `dr_cb`.

Opinions anyone? 
@webmakersteve ?

**Comment by mimaison** [comment]
As @edoardocomar has shown, the throughput difference is negligible. There might be a slight increase in memory pressure (again only with delivery report enabled) as the buffer is kept in memory until the report is received. 

I'm not very familiar with the benchmark used, but I'd be curious to see how the results change if we have large keys. We'll give that a try.

While it's true librdkafka does not have this automatic opaque behaviour, it still provides the payload in delivery report simply by enabling dr_msg_cb (which was the mechanism we introduced in #72 but it was removed when the opaque was added).

Having the payload/key available in the report is a **key** feature. It enables the application to resend messages that failed.

I can understand not everybody might want this automatic opaque but I think it makes sense to have it if dr_msg_cb is enabled.

**Comment by edoardocomar** [comment]
PR updated - please review

**Comment by webmakersteve** [comment]
I think I'd prefer implementing `dr_msg_cb` on the c++ side for this use case to make delivery reports automatically include the message in the report.

I've already begun an implementation that will satisfy this. I should have it finalized in the next few days.

**Comment by webmakersteve** [comment]
Fixed with #113
